src/main/java/org/jabref/gui/preferences/TableColumnsTab.java

```diff
@@ -96,81 +96,80 @@
     public TableColumnsTab(JabRefPreferences prefs, JabRefFrame frame) {
         this.prefs = prefs;
         this.frame = frame;
-        this.data = FXCollections.observableArrayList(
-                new TableRow("entrytype",75),
-                new TableRow("author/editor",300),
-                new TableRow("title",470),
-                new TableRow("year",60),
-                new TableRow("journal",130),
-                new TableRow("bibtexkey",100));
 
+        /* Populate the data of Entry table columns */
+        List<String> prefColNames = this.prefs.getStringList(this.prefs.COLUMN_NAMES);
+        this.data = FXCollections.observableArrayList();
+        for (int i = 0; i < prefColNames.size(); i++) {
+            this.data.add(new TableRow(prefColNames.get(i)));
+        }
+
+        /* UI for Entry table columns */
         colSetup = new TableView<>();
-        TableColumn<TableRow,String> field = new TableColumn<>(Localization.lang("Field name"));
-        TableColumn<TableRow,Double> column = new TableColumn<>(Localization.lang("Column width"));
+        colSetup.setEditable(true);
+        TableColumn<TableRow, String> field = new TableColumn<>(Localization.lang("Field name"));
         field.setPrefWidth(400);
-        column.setPrefWidth(240);
         field.setCellValueFactory(new PropertyValueFactory<>("name"));
         field.setCellFactory(TextFieldTableCell.forTableColumn());
+        field.setEditable(true);
         field.setOnEditCommit(
                 (TableColumn.CellEditEvent<TableRow, String> t) -> {
                     t.getTableView().getItems().get(
                             t.getTablePosition().getRow()).setName(t.getNewValue());
-                });
-        column.setCellValueFactory(new PropertyValueFactory<>("length"));
-        column.setOnEditCommit(
-                (TableColumn.CellEditEvent<TableRow, Double> t) -> {
-                    t.getTableView().getItems().get(
-                            t.getTablePosition().getRow()).setLength(t.getNewValue());
+                    // Since data is an ObservableList, updating it updates the displayed field name.
+                    this.data.set(t.getTablePosition().getRow(), new TableRow(t.getNewValue()));
+                    // Update the User Preference of COLUMN_NAMES
+                    List<String> tempColumnNames = this.prefs.getStringList(this.prefs.COLUMN_NAMES);
+                    tempColumnNames.set(t.getTablePosition().getRow(), t.getNewValue());
+                    this.prefs.putStringList(this.prefs.COLUMN_NAMES,tempColumnNames);
                 });
 
+        tableRows.clear();
```

Comment:
Do you understand the purpose of the `tableRows` field? It appears that it is always synced to the `data` collection anyway. If that's the case, could you please remove `tableRows`. Thanks.

Reply:
The tableRow is not the Observable Array List (that's the data field), it's there to make sure that changes in the order and number of field names are synced with changes in the column widths. The data collection is an Observable Array List that only contains the field names, so without tableRows the changes won't be saved.